Mathlib/Topology/Algebra/Valued/WithZeroMulInt.lean

```diff
@@ -45,4 +44,54 @@ lemma exists_pow_lt_of_le_exp_neg_one [Valued R ℤᵐ⁰] {x : R} (hx : v x ≤
   rw [← exp_zero, exp_lt_exp]
   norm_num
 
+variable {K : Type*} [Field K] [Valued K ℤᵐ⁰]
+
+theorem irreducible_valuation_lt_one {ϖ : 𝒪[K]} (h : Irreducible ϖ) : v ϖ.1 < 1 :=
+  lt_of_le_of_ne (Valuation.mem_integer_iff _ _ |>.1 ϖ.2) <|
+    mt (Valuation.integer.integers _).isUnit_iff_valuation_eq_one.2 h.not_isUnit
+
+theorem irreducible_valuation_le_exp_neg_one {ϖ : 𝒪[K]} (h : Irreducible ϖ) :
+    v ϖ.1 ≤ exp (-1 : ℤ) := by
+  have hϖ : v ϖ.1 ≠ 0 := by simp [h.ne_zero]
+  have := log_one (M := ℤ) ▸  (log_lt_log hϖ one_ne_zero).2 (irreducible_valuation_lt_one h)
+  rw [← log_le_iff_le_exp hϖ]
+  linarith
+
+theorem mem_maximalIdeal_pow_valuation [IsDiscreteValuationRing 𝒪[K]]
+    {x : 𝒪[K]} {n : ℕ} (hx : x ∈ 𝓂[K] ^ n) {ϖ : 𝒪[K]} (h : Irreducible ϖ) :
+    v x.val ≤ v ϖ.1 ^ n := by
+  rcases eq_or_ne x 0 with (rfl | hx₀) <;> try simp
+  simp_rw [h.maximalIdeal_eq, Ideal.span_singleton_pow, Ideal.mem_span_singleton] at hx
+  obtain ⟨y, rfl⟩ := hx
+  simpa using le_trans (mul_le_of_le_one_right' <| (Valuation.mem_integer_iff _ _).1 y.2) le_rfl
+
+/-- The ring of integers `𝒪[K]` of a `ℤₘ₀`-valued field `K` with finite residue
+field has a finite covering by elements of the basis of uniformity of `K`, whenever
+`𝒪[K]` is a discrete valuation ring. -/
+theorem finite_cover_of_uniformity_basis [IsDiscreteValuationRing 𝒪[K]] {γ : ℤᵐ⁰ˣ}
+    (h : Finite 𝓀[K]) :
+    ∃ t : Set K, Set.Finite t ∧
+      (𝒪[K]).carrier ⊆ ⋃ y ∈ t, { x | (x, y) ∈ { p | v (p.2 - p.1) < γ.val } } := by
+  classical
+  let ⟨ϖ, hϖ⟩ := IsDiscreteValuationRing.exists_irreducible 𝒪[K]
+  let ⟨m, hm⟩ := exists_pow_lt_of_le_exp_neg_one (irreducible_valuation_le_exp_neg_one hϖ) γ
+  have := integer.finite_quotient_maximalIdeal_pow_of_finite_residueField h m
+  have h := Fintype.ofFinite (𝒪[K] ⧸ 𝓂[K] ^ m)
+  let T := Subtype.val '' (h.elems.image Quotient.out).toSet
+  refine ⟨T, (Set.Finite.image _ (Finset.finite_toSet _)), fun x hx ↦ ?_⟩
+  simp only [Set.mem_iUnion]
+  let y := (Ideal.Quotient.mk (𝓂[K] ^ m) ⟨x, hx⟩).out
+  exact ⟨y, Set.mem_image_of_mem _ <| Finset.mem_image_of_mem _ (h.complete _),
+    lt_of_le_of_lt (mem_maximalIdeal_pow_valuation (Ideal.Quotient.out_sub _ _) hϖ) hm⟩
+
+variable (K)
+
+/-- The ring of integers `𝒪[K]` of a complete `ℤₘ₀`-valued field `K` with finite residue
+field is compact, whenever `𝒪[K]` is a discrete valuation ring. -/
+theorem integer_compactSpace [CompleteSpace K] [IsDiscreteValuationRing 𝒪[K]] (h : Finite 𝓀[K]) :
+    CompactSpace 𝒪[K] where
+   isCompact_univ := isCompact_iff_isCompact_univ.1 <| isCompact_iff_totallyBounded_isComplete.2
+      ⟨(hasBasis_uniformity _ _).totallyBounded_iff.2 fun _ _ ↦ finite_cover_of_uniformity_basis h,
+        (isClosed_integer K).isComplete⟩
```

Comment:
Why can't you use `Valued.integer.properSpace_iff_completeSpace_and_isDiscreteValuationRing_integer_and_finite_residueField` directly? Instead of having to re-prove the cover argument, which is already inside `Valued.integer.totallyBounded_iff_finite_residueField`?